lib/connections/limiter.go

```diff
@@ -12,53 +12,126 @@ import (
 	"sync/atomic"
 
 	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
 	"golang.org/x/net/context"
 	"golang.org/x/time/rate"
+	"sync"
 )
 
 // limiter manages a read and write rate limit, reacting to config changes
 // as appropriate.
 type limiter struct {
-	write     *rate.Limiter
-	read      *rate.Limiter
-	limitsLAN atomicBool
+	write               *rate.Limiter
+	read                *rate.Limiter
+	limitsLAN           atomicBool
+	deviceReadLimiters  *sync.Map
+	deviceWriteLimiters *sync.Map
```

Comment:
`sync.Map` is Go 1.9 and we're currently targeting 1.8. Although we *could* bump our requirements I think there should be little enough churn on these that we can actually use regular maps and locking...